### PR TITLE
COCO Detection Dataset Import/Export support

### PR DESCRIPTION
**Pull request by hardikdava**
# Description

- Added a support for `COCO` detection dataset import and export
- Added a support for reading and writing `json` files
- Related issue #141 

## Type of change

-   [x] New feature (non-breaking change which adds functionality)
-   [x] This change requires a documentation update

## How has this change been tested?

- Tested locally with dataset coco128 (downloaded from roboflow)
- Import using `supervision.DetectionDataset`
- Visualized using `supervision.BoxAnnotator`
- Saved dataset using `supervision.DetectionDataset`

## Docs

-   [x] Docs updated?

### CODE REVIEW

**Comment by yclicc** [comment]
Is this supposed to support segmentation masks? It doesn't seem to be doing that for me, it only exports the bounding boxes.

**Comment by hardikdava** [comment]
@yclicc Not yet. But it will be supported in near future. Currently, it is supporting only object detection format.

**Comment by SkalskiP** [comment]
Hi @hardikdava 👋🏻 ! I was AFK at the CVPR conference last week. I'll do a code review today.

**Comment by hardikdava** [comment]
@SkalskiP I also added support for instance segmentation masks loading. I tested locally. Please also test with coco segmentation dataset.

**Comment by hardikdava** [comment]
> Is this supposed to support segmentation masks? It doesn't seem to be doing that for me, it only exports the bounding boxes.

@yclicc You can try now. It should be working.

**Comment by yclicc** [comment]
@hardikdava that throws an error with my data on line 84 of dataset/formats/coco.py because "_polygons" is a dict. COCO format allows for segmentations either as a list of polygons or as a run length encoding (which is what I've got).

image

**Comment by hardikdava** [comment]
@yclicc , have you used latest code? please share your dataset if possible? I tried using segmentation dataset from roboflow.

**Comment by yclicc** [comment]
This is latest code, but your code is expecting a list of polygons (which is a possible segmentation format for COCO, just not the only option). See https://opencv.org/blog/2021/10/12/introduction-to-the-coco-dataset/ for details and examples of both possible types of mask format.

**Comment by yclicc** [comment]
I can't share my dataset, but I'm generating it with CVAT's mask annotation mode, as opposed to the polygon creator. So perhaps unsurprisingly, it has exported a COCO format with Run Length Encoding instead of a list of polygons.

**Comment by hardikdava** [comment]
@yclicc this PR only supports for polygons. But I can take a look and check for possibility to extend the functionality.

**Comment by yclicc** [comment]
I'll see if I can come up with a fix and then put in a pull request on your repo.

**Comment by SkalskiP** [comment]
@hardikdava could you also run those two commands below in the root of your fork directory? This will ensure your new code is formatted constantly with rest of `supervision`.

```
black supervision
isort --profile black supervision/
```

File: supervision/dataset/formats/coco.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        ],
+        dtype=bool,
+    )
+
+
+def coco_annotations_to_detections(image_annotations: List[dict], resolution_wh: Tuple[int, int], with_masks: bool) \
+        -> Detections:
+    """
+    Returns:
+        object: detections
```
Comment (SkalskiP) [inline supervision/dataset/formats/coco.py:71]: Let's throw out this comment. It is redundant we see the type.

File: supervision/utils/file.py [hunk 2]
```diff
@@ -66,3 +81,31 @@ def save_text_file(lines: List[str], file_path: str):
...
+
+    Returns:
+        dict: A dict of annotations information
+    """
+    with open(file_path, "r") as file:
+        data = json.load(file)
+    return data
+
+
+def save_json_file(data: dict, file_path: str, indent=3):
```
Comment (SkalskiP) [inline supervision/utils/file.py:101]: Let's add type to `indent` argument. And add the return type `None`.

File: supervision/utils/file.py [hunk 3]
```diff
@@ -1,6 +1,21 @@
+import json
 from pathlib import Path
 from typing import List, Optional, Union
 
+import numpy as np
+
+
+class NpEncoder(json.JSONEncoder):
```
Comment (SkalskiP) [inline supervision/utils/file.py:8]: Could we rename it to `NpEncoder`? I prefer full names whenever possible.

**Review by SkalskiP — CHANGES_REQUESTED** [review]

**Review by SkalskiP — CHANGES_REQUESTED** [review]

File: supervision/dataset/core.py [hunk 4]
```diff
@@ -333,6 +339,102 @@ def as_yolo(
...
+            >>> roboflow.login()
+
+            >>> rf = Roboflow()
+
+            >>> project = rf.workspace(WORKSPACE_ID).project(PROJECT_ID)
+            >>> dataset = project.version(PROJECT_VERSION).download("coco")
+
+            >>> ds = sv.DetectionDataset.from_coco(
+            ...     images_directory_path=f"{dataset.location}/train/images",
+            ...     annotations_directory_path=f"{dataset.location}/train/annotations.json",
```
Comment (SkalskiP) [inline supervision/dataset/core.py:375]: Incorrect argument name in docs -`annotations_directory_path`. Inconsistent with method argument name -`annotations_path`.

**Review by SkalskiP — CHANGES_REQUESTED** [review]

File: supervision/dataset/core.py (hunk 4 above)
Comment (SkalskiP) [reply supervision/dataset/core.py:375]: Also Roboflow use a bit different naming convention so it is not `annotations.json` but `_annotations.coco.json`.

**Review by SkalskiP — CHANGES_REQUESTED** [review]

File: supervision/dataset/formats/coco.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+    annotations_data = []
+    image_infos = []
+    categories = []
+
+    for class_id, class_name in enumerate(classes):
+        cate_dict = {'id': class_id, 'name': class_name, "supercategory": "common-objects"}
+        categories.append(cate_dict)
+
+    image_id = 0
```
[the remark this answers is not in the document]
Comment (hardikdava) [reply supervision/dataset/formats/coco.py:176]: @SkalskiP This implementation is done without pycocotools and I found `image_id` and `label_id` starting with `0` from roboflow dataset.

File: supervision/dataset/formats/coco.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        annotation = coco_annotations_to_detections(image_annotations=image_annotations, resolution_wh=(w, h),
+                                                    with_masks=force_masks)
+
+        images[images_info["file_name"]] = image
+        annotations[images_info["file_name"]] = annotation
+
+    return classes, images, annotations
+
+
+def save_coco_annotations(
```
[the remark this answers is not in the document]
Comment (hardikdava) [reply supervision/dataset/formats/coco.py:151]: @SkalskiP I am breaking long method into shorter. Give me some time, please.

File: supervision/dataset/formats/coco.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SkalskiP) [reply supervision/dataset/formats/coco.py:176]: > I found `image_id` and `label_id` starting with `0` from roboflow dataset.

If so then that's good enough due diligence for me.

**Comment by hardikdava** [comment]
@SkalskiP I have committed the changes you requested. Also ran two commands with `black` and `isort`. Please review it.

File: supervision/dataset/formats/coco.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+import cv2
+import numpy as np
+
+from supervision.dataset.ultils import approximate_mask_with_polygons
+from supervision.detection.core import Detections
+from supervision.detection.utils import polygon_to_mask
+from supervision.utils.file import read_json_file, save_json_file
+
+
+def _extract_class_names(annotation_data: dict) -> List[str]:
```
Comment (SkalskiP) [inline supervision/dataset/formats/coco.py:15]: I noticed a bug in this method. It does not work with the hierarchy of classes. Could you take a look at this example?

```json
[
    {
        "id": 0,
        "name": "fashion-assistant",
        "supercategory": "none"
    },
    {
        "id": 1,
        "name": "baseball cap",
        "supercategory": "fashion-assistant"
    },
    {
        "id": 2,
        "name": "hoodie",
        "supercategory": "fashion-assistant"
    },
    {
        "id": 3,
        "name": "jacket",
        "supercategory": "fashion-assistant"
    },
    {
        "id": 4,
        "name": "pants",
        "supercategory": "fashion-assistant"
    },
    {
        "id": 5,
        "name": "shirt",
        "supercategory": "fashion-assistant"
    },
    {
        "id": 6,
        "name": "shorts",
        "supercategory": "fashion-assistant"
    },
    {
        "id": 7,
        "name": "sneaker",
        "supercategory": "fashion-assistant"
    },
    {
        "id": 8,
        "name": "sunglasses",
        "supercategory": "fashion-assistant"
    },
    {
        "id": 9,
        "name": "sweatshirt",
        "supercategory": "fashion-assistant"
    },
    {
        "id": 10,
        "name": "t-shirt",
        "supercategory": "fashion-assistant"
    }
]
```

Here `fashion-assistant` is a top category - parent to other categories. It should be discarded during loading. So it should be loaded as:

```
[
    'baseball cap',
    'hoodie',
    'jacket',
    'pants',
    'shirt',
    'shorts',
    'sneaker',
    'sunglasses',
    'sweatshirt',
    't-shirt'
]
```

Also, I'm not 100 sure if this implementation guarantee loading class names in the order of class ids.

Comment (SkalskiP) [reply supervision/dataset/formats/coco.py:15]: I would even use that example as a case for the unit test.

File: supervision/dataset/formats/coco.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+    image_id_label_id_pair, annotation_dict = _annotations_dict(
+        annotation_data=annotation_data
+    )
+
+    images = {}
+    annotations = {}
+
+    for images_info in images_infos:
+        image_path = os.path.join(images_directory_path, images_info["file_name"])
+        image = cv2.imread(str(image_path))
```
Comment (SkalskiP) [inline supervision/dataset/formats/coco.py:172]: Just tested dataset loading and noticed that images were not loaded correctly.

Here is my code:

```python
import roboflow
from roboflow import Roboflow

roboflow.login()

rf = Roboflow()
project = rf.workspace("roboflow-jvuqo").project("fashion-assistant-segmentation")
dataset = project.version(5).download("coco")

import supervision as sv

ds = sv.DetectionDataset.from_coco(
     images_directory_path=f"{dataset.location}/train/images",
     annotations_path=f"{dataset.location}/train/_annotations.coco.json",
)
```

And images got loaded as:

```
...
'photo_2023-05-30-17-34-54_jpeg.rf.593fa7603782b4679c624d5c1f63a508.jpg': None,
 'photo_2023-05-29-12-28-55_jpeg_jpg.rf.5a577cabb1aa09018e06fe01640bbb03.jpg': None,
 'photo_2023-05-29-12-29-08_jpeg_jpg.rf.5bd8bfb57cd46e45f2fd8ff9bfa34de1.jpg': None,
 'photo_2023-05-29-14-10-23_jpeg_jpg.rf.5d4449e52e9dcdf25ef1f5b3f1c1b62c.jpg': None,
 'photo_2023-05-30-17-35-05-1-_jpeg.rf.5c217edf22eab45f1761f3972453d469.jpg': None,
 'photo_2023-05-30-17-34-57_jpeg.rf.5959091a468b991b5794778f093c881a.jpg': None,
 'photo_2023-05-29-12-29-12_jpeg_jpg.rf.5b0744781d8f14023a55a9edadfc23b1.jpg': None,
...
```

**Review by SkalskiP — CHANGES_REQUESTED** [review]

**Comment by SkalskiP** [comment]
@hardikdava I did some tests and noticed two critical bugs. I marked them in the code. You can also use this Google Colab as reference: https://colab.research.google.com/drive/1spwwYyYO-LN3RDzp5s8whG1_7JzROm-m?usp=sharing

**Comment by SkalskiP** [comment]
@hardikdava 

> I would like to work on new features. I know unittests are important but I am not fan of implementing them ;)

No worries I can take care of that. For now, let's make sure simple tests in Google Colab work. And when we will be done with that. I'll make unittests ;)

**Comment by SkalskiP** [comment]
@hardikdava I've done that already :) Also, do I have the right to commit to your fork? I could start to work on unit tests on your branch.

File: supervision/dataset/formats/coco.py (hunk 7 above)
Comment (SkalskiP) [reply supervision/dataset/formats/coco.py]: @hardikdava I solved that problem with https://github.com/roboflow/supervision/pull/150/commits/540efc80ce2cfcd37539d35a621575db1f536764

**Comment by SkalskiP** [comment]
Hi @hardikdava 👋🏻 ! I spend a few hours today writing unit tests and code re-factorication. I'm not done yet, but I hope I'll finish tomorrow. So far so good.

**Comment by hardikdava** [comment]
Hi @SkalskiP 👋🏻 , I checked your commits. I have to adopt coding style to supervision format. Let me know if I can help you in anyway in progress. I would like to keep continue contributions.

**Comment by SkalskiP** [comment]
> I have to adopt coding style to supervision format.

No worries :) I try to keep API and style consistent.

> Let me know if I can help you in anyway in progress. I would like to keep continue contributions.

For now, my work comes down primarily to writing unit tests and cleaning up here. As you try to stay away from testing, I think I can take it from here, but we can think about the next task for you.

**Review by SkalskiP — APPROVED** [review]

**Comment by SkalskiP** [comment]
Hi @hardikdava 👋🏻 ! We are ready for merge ;)

**Comment by hardikdava** [comment]
Hi @SkalskiP 👋🏻, thanks for merging my contribution. I am glad to see my contribution in community. Always happy to help! how should we discuss about new feature in github or else?

**Comment by SkalskiP** [comment]
@hardikdava 🔥 

- I'm still working on the description of the API for: https://github.com/roboflow/supervision/discussions/153. I want to do it right so it will probably take me a while to come up with something I'd be happy with.
- Would you like to take care of: https://github.com/roboflow/supervision/issues/155?

**Comment by yclicc** [comment]
I've found a problem with this code, in that the image_annotation["category_id"] might be indexed from 1 instead of from zero, or may even skip numbers (not sure if that is valid COCO, but it should be coped with properly). However, coco_categories_to_classes returns the class names as an (obviously zero indexed) list with no skips, but then coco_annotations_to_detections sets the class_ids to simply the image_annotation["category_id"] even though this may not be zero indexed.

**Comment by SkalskiP** [comment]
Hi @yclicc, I noticed that problem while reviewing one of the current PRs. The fix should be in `main` now. Could you try using it once again?

**Comment by yclicc** [comment]
Yep, that's fixed it. Thanks!

**Comment by SkalskiP** [comment]
@yclicc awesome! We will probably release those changes today or tomorrow.

**Comment by yclicc** [comment]
Ah, no I'm afraid this is still present. Suggest modifying coco_categories_to_classes to make it also return a lookup table from category["id"] to index in the new category list.

**Comment by SkalskiP** [comment]
I just tested it. It works for me. Making it a lookup table would require changing every other format we support. Can you give me any example of a dataset it doesn't work for?

**Comment by yclicc** [comment]
Any COCO dataset where the numbering of the ids of the categories doesn't start at 0 and/or skips a number.

E.g. ```"categories": [{"id":1, "name":"Albatross", "supercategory": ""}, {"id": 3,"name":"Walrus", "supercategory":""}]```
That will then return a dataset.yaml of the form
```
names:
- Albatross
- Walrus
nc: 2
```
but the indices of each row of the YOLO dataset will be 1 and 3 instead of 0 and 1. By a lookup table, you don't need to modify anything except the COCO code. Have coco_categories_to_classes return a lookup table, then have coco_annotations_to_detections accept that lookup table and then when the class_ids variable is set have it lookup in that lookup table for each `image_annotation["category_id"]`. The only downside is that converting from COCO to itself will lose information by changing the IDs in the outputted COCO. But I'd say that's a price worth paying to have the result work properly with YOLO.

**Comment by SkalskiP** [comment]
So the ids of your input dataset are not ordered from `0` to `n-1` where `n` is the number of categories? I see how it breaks. 

In that case, we have two options:
- Remapping ids during COCO loading (I would prepare that)
- Storing classes as dictionary not a list

**Comment by SkalskiP** [comment]
@yclicc Yeah this is something that we need to take into consideration :/

**Comment by SkalskiP** [comment]
@hardikdava could you take a look at the problem @yclicc described?

**Comment by hardikdava** [comment]
@SkalskiP sure. I will take a look at it and try to fix it.

**Comment by SkalskiP** [comment]
@hardikdava awesome!

**Comment by hardikdava** [comment]
@SkalskiP and @yclicc some of the things I found about coco dataset.


**Issue: Category map ids assignment**
- COCO dataset category map ids starting from `1` not `0`. Source:  [src1](https://www.v7labs.com/blog/coco-dataset-guide) [src2](https://www.immersivelimit.com/tutorials/create-coco-annotations-from-scratch) [src3](https://opencv.org/blog/2021/10/12/introduction-to-the-coco-dataset/)
- But coco dataset downloaded from Robfolow has category map ids starting from `0`.
- Even though starting ids from `0` does not create an issue. Tested with pycocotools(python API to work with coco dataset).
- We can make it something like ids will start from `1` for only coco dataset.

**Issue: Segmentation polygon type**
- I only found the below suggested format by [opencv university](https://opencv.org/blog/2021/10/12/introduction-to-the-coco-dataset/) and @yclicc 
```
“annotations”: [{
    ”segmentation”:
    {	
        “counts”: [34, 55, 10, 71]
        “size”: [240, 480]
    },
    “area”: 600.4,
    “iscrowd”: 1,
    “Image_id:” 122214,
    “bbox”: [473.05, 395.45, 38.65, 28.92],
    “category_id”: 15,
    “id”: 934
}
```
- I personally download instance segmentation label from CVAT and it does not have above mentioned format.
- I would suggest let the segmentation as polygons and later if we receive some more issue from other users then we will plan to improve it. 

Let me know your views soon.

**Comment by yclicc** [comment]
**Category map ids assignment**
The thing is, once it is loaded as a detection we want it to be agnostic to the fact that came from COCO format. So you would want to make it zero indexed so that when it is then written to YOLO (the most common usecase for the Supervision library I would think?) it works properly and plays nicely with the YOLO models. If indeed it is the _convention_ to start at 1 for COCO then the exporter that goes back to COCO could of course do that.

**Segmentation polygon type**
It remains the case that the RLE format is part of the COCO spec, though, so you may assume that people will try to load such data in. As it is now, it throws an unhandled exception when this happens. Might it be worth at least doing a try catch and raising a NotImplementedError if segmentation is a dict instead of a numpy array?

**Comment by hardikdava** [comment]
@yclicc my thoughts,

**Segmentation polygon type**
I can add something like if the segmentation annotation datatype is `dict` then it will raise `NotImplementedError`. This can easily be done.

**Category map ids assignment**
- `supervision` is not bounded by `yolo`. So it has to be work with `yolo` or any other format.
- The idea of offset will be included while writing the coco dataset. But currently, `supervision` does not have any information whether the original dataset is loaded from `coco` or `yolo` so think of following situations:
- situation-1: dataset is loaded from `coco` and saving in `coco` format then offset of `1in class_id` is not needed. Because `class_id` is already starting from `1`
- situation-2: dataset is loaded from other than `coco` format and saving in `coco` format then offset of `1 in class_id` is needed as `coco` `class_ids` are starting from `1`.

**Comment by SkalskiP** [comment]
Hi @hardikdava and @yclicc 👋🏻 How about we just remap class ids when we load COCO annotations?

**Comment by hardikdava** [comment]
@yclicc @SkalskiP Found the solution. We should offset the `coco-category_id` by `-1` when we load the coco dataset. So every `class ids` are starting from `0` regardless of dataset. We can offset `class id` by `+1` while we are storing dataset only for `coco` format. This will fix the issue of storing the dataset. I do not see any issue for this.

**Comment by hardikdava** [comment]
@yclicc can you try with #169 ? Let me know how it goes.

**Comment by SkalskiP** [comment]
@yclicc I created a fix for this issue https://github.com/roboflow/supervision/pull/176. Could you take a look and try to see if it works?

**Comment by yclicc** [comment]
Seems good to me, though it still fails as my dataset contains RLEs rather than polygons.

**Comment by yclicc** [comment]
Yep, all seems to work fine when I add my RLE code back in (which sadly I can't share). Thanks!

**Comment by SkalskiP** [comment]
> Yep, all seems to work fine when I add my RLE code back in (which sadly I can't share). Thanks! 

@yclicc We will add `RLE` support in feature releases 💯 Thanks for testing. 

@hardikdava looks like we will move forward with https://github.com/roboflow/supervision/pull/176.

**Pull request merged**